extensions/resource-deployment/src/ui/validation/validations.ts

```diff
@@ -0,0 +1,227 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the Source EULA. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+import * as azdata from 'azdata';
+import { EOL } from 'os';
+import { throwUnless } from '../../common/utils';
+import * as loc from '../../localizedConstants';
+import { InputComponent } from '../modelViewUtils';
+import { RadioGroupLoadingComponentBuilder } from '../radioGroupLoadingComponentBuilder';
+
+export interface ValidationState {
+	valid: boolean;
+	message?: string;
+}
+
+export type Validator = () => Promise<ValidationState>;
+export type VariableValueGetter = (variable: string) => (Promise<string | number | undefined | null> | Promise<string | undefined | null>);
+export type ValueGetter = () => (Promise<string | number | undefined | null> | Promise<string | undefined | null>);
+
+export const enum ValidationType {
+	IsInteger = 'is_integer',
+	Regex = 'regex_match',
+	LessThanOrEquals = '<=',
+	GreaterThanOrEquals = '>='
+}
+
+export type IValidation = IRegexValidation | IIntegerValidation | IComparisonValidation;
+
+export interface IValidationBase {
+	readonly type: ValidationType,
+	readonly description: string,
+	getValidator?(): Validator
+}
+
+export type IIntegerValidation = IValidationBase;
+
+export interface IRegexValidation extends IValidationBase {
+	readonly regex: string | RegExp
+}
+
+export interface IComparisonValidation extends IValidationBase {
+	readonly target: string
+}
+
+export abstract class Validation implements IValidationBase {
+	private _type: ValidationType;
+	private _description: string;
+
+	get type(): ValidationType {
+		return this._type;
+	}
+
+	get description(): string {
+		return this._description;
+	}
+
+	// gets the validator for this validation object
+	abstract getValidator(): Validator;
+
+	protected getValue(): Promise<string | number | undefined | null> {
+		return this._valueGetter();
+	}
+
+	protected getVariableValue(variable: string): Promise<string | number | undefined | null> {
+		return this._variableValueGetter!(variable);
+	}
+
+	constructor(validation: IValidation, protected _valueGetter: ValueGetter, protected _variableValueGetter?: VariableValueGetter) {
```

Comment:
So we pass in an IValidation, but the class itself is also an IValidation? And then we don't do anything with the getValidator that's passed in? That seems very odd - why not just have the input be their own params/options?

Reply:
I am not sure what you mean by in put having its own params/options?
The key difference between IValidation and Validation is the getValidator, which is always undefined for IValidation and defined for Validaiton. If there is a better way to set this up, I would love it. Initially I did not have getValidator() in IValidation at all but then somewhere I needed it, I will see if I can remove it again as per my original intention.

Reply:
If it's always going to be undefined then why even have it be a part of the interface is then the question I'd have. You can make an extra interface that extends the IValidation base one to add in the getValidator if you really need to use the interface.

And what I'm saying is there a reason you're taking in an IValidation as opposed to just having parameters for type and description in the constructor?

Reply:
> @ranasaria the input field type already has the number type, and you can specify min and max, are you aware of that? it is already exposed and available for you to use in JSON.

Yes I know about them. These less than or greater than are meant to do comparisons with other field values. That is why the 'target' string which is a field variable exists.

Reply:
So IValidation just meant to be the input object to the Validation object. I guess I should call it ValidationInfo to avoid the confusion. I will remove the fact that Validation implements IValidation.

Reply:
Addressed